staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go

```diff
@@ -367,15 +366,16 @@ func newTestableController(config TestableConfig) *configController {
 	return cfgCtlr
 }
 
-func (cfgCtlr *configController) Run(ctx context.Context) error {
-	defer utilruntime.HandleCrash()
-
-	// Let the config worker stop when we are done
-	defer cfgCtlr.configQueue.ShutDown()
+func (cfgCtlr *configController) Start(ctx context.Context) error {
+	go func() {
+		<-ctx.Done()
+		// Let the config worker stop when we are done
+		defer cfgCtlr.configQueue.ShutDown()
+	}()
```

Comment:
since we are here, it would be nice if there is a way for the caller to wait for the config controller and all it child goroutines to be done, it helps the tests to be flake free.
```
func (cfgCtlr *configController) Start(ctx context.Context) (done context.Context, errStart error)
{
      if ok := cache.WaitForCacheSync(ctx.Done(), cfgCtlr.plInformerSynced, cfgCtlr.fsInformerSynced); !ok {
           return nil, fmt.Errorf("never achieved initial sync")
       }

       var wg sync.WaitGroup
       wg.Add(2)

       done, cancel := context.WithCancelCause(context.Background())
       go func() {
		<-ctx.Done()
             
		// Let the config worker stop when we are done
		cfgCtlr.configQueue.ShutDown() 
                wg.Wait()

                 cancel(nil)
	}() 
        
        go func() {
               defer wg.Done()
                wait.UntilWithContext(ctx, cfgCtlr.runWorker, time.Second)                    
        }()

        go func() {
               defer wg.Done()
                wait.UntilWithContext(ctx, cfgCtlr.updateBorrowing, borrowingAdjustmentPeriod)
        }()
      
        return done, nil 
}
```
 
The post start hook does not need to use the `done` ctx, it only needs to check `errStart`

Reply:
This hasn't been the case before either, has it? The old `configQueue.Shutdown()` did not block.

Reply:
@sttts: I do not understand how your comment relates to @tkashem's suggestion. His point is to introduce a way for the caller to know when the forked goroutines that process the workqueue items and and periodically rebalance the concurrency allocations have completed. He suggests adding a goroutine that waits for both of the others to be done and then cancels the done context to signal that joint completion.

Reply:
This is out of scope of this PR. Am not arguing against it. It would be new functionality and deserves to be a follow-up.

Reply:
OK, that makes sense to me.

Reply:
for the follow up: use `ShutDownWithDrain` instead of `configQueue.Shutdown()`

Reply:
Blocking `Run` is always a little awkward because you cannot distinguish between "it never started" and "it started but later returned an error". I like the idea of `Start` with an additional `done` channel better. It's more explicit and flexible.

Reply:
Or in other words, my made up Go proverb: leave the control flow to the caller.

Reply:
In this case, however, the errors only arise during startup?

Reply:
It was only `return fmt.Errorf("never achieved initial sync")`. All other paths returned nil.

Reply:
Let's pursue subtleties of the follow-up in the follow-up.